pkg/apis/gcpfirewall/v1alpha1/types.go

```diff
@@ -0,0 +1,186 @@
+// secure tags) which are deduced by the controller. As a result, the specified
+// rule applies to ALL nodes and pods in the cluster.
+type GCPFirewallSpec struct {
+	// Rule action of the firewall rule. Only allow action is supported. If not
+	// specified, defaults to ALLOW.
+	// +optional
+	// +kubebuilder:validation:Enum=ALLOW
+	// +kubebuilder:default=ALLOW
+	Action Action `json:"action"`
+
+	// List of protocol/ ports which needs to be selected by this rule.
+	// If this field is empty or missing, this rule matches all protocol/ ports.
+	// If this field is present and contains at least one item, then this rule
+	// allows traffic only if the traffic matches at least one port in the list.
+	// +optional
+	Ports []ProtocolPort `json:"ports,omitempty"`
+
+	// List of sources that are allowed by this rule. Items in this list are
+	// combined using a logical OR operation. If this field is empty or missing,
+	// this rule allows all sources. If this field is present and contains at
+	// least one item, this rule allows traffic only if the traffic matches at
+	// least one item in the from list.
+	// +optional
+	Ingress []GCPFirewallIngressPeer `json:"ingress,omitempty"`
+}
+
+// GCPFirewallIngressPeer describes a peer to allow traffic from.
+type GCPFirewallIngressPeer struct {
+	// IPBlocks specify the set of CIDRs that the rule applies to.
+	// Valid example list items are "192.168.1.1/24" or "2001:db9::/64".
+	// +optional
+	// +kubebuilder:validation:MaxItems=256
+	IPBlocks []CIDR `json:"ipBlocks,omitempty"`
+}
+
+// ProtocolPort describes the protocol and ports to allow traffic on.
+type ProtocolPort struct {
+	// The protocol which the traffic must match.
+	// +kubebuilder:validation:Enum=TCP;UDP;ICMP;SCTP;AH;ESP
+	Protocol Protocol `json:"protocol"`
+
+	// StartPort is the starting port of the port range that is selected on the
+	// firewall rule targets for the specified protocol. If EndPort is not
+	// specified, this is the only port selected.
+	// If StartPort is not provided, all ports are matched.
+	// +optional
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	StartPort *int32 `json:"startPort,omitempty"`
+
+	// EndPort is the last port of the port range that is selected on the firewall
+	// rule targets. If StartPort is not specified or greater than this value, then
+	// this field is ignored.
+	// +optional
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	EndPort *int32 `json:"endPort,omitempty"`
+}
+
+// GCPFirewallStatus is the runtime status of a GCP firewall
+type GCPFirewallStatus struct {
+	// Type specifies the underlying GCE firewall implementation type.
+	// Takes one of the values from [VPC, REGIONAL, GLOBAL]
+	// +optional
+	// +kubebuilder:validation:Enum=VPC;REGIONAL;GLOBAL
+	Type string `json:"type,omitempty"`
+
+	// Name of the GCP firewall rule in the GCE.
+	// +optional
+	Name string `json:"name"`
```

Comment:
I am wondering if you should just include the selflink/URL of the firewall object directly. 

No need to have a specific name

Reply:
This would remove the need for specifying global vs. region as well.

Reply:
Do we want to include project name/ number in the status ?. Also, the prefix googleapis.com keeps changing which might make the status uncertain. 

We have region vs global concept only for FW 3.0. With FW 3.0, firewall rules are not first class citizens in GCE anymore so they do not have self links. All GKE cluster rules will be a part of single firewall network policy resource.

A VPC firewall status will look like:
```
status:
  type: VPC
  name: gke-pfw-default-my-ingress
  conditions:
  - type: Enforced
     status: True
```

A global network firewall policy status will be:
```
status:
  type: GLOBAL
  priority: 10
  conditions:
  - type: Enforced
     status: True
```

A regional network firewall policy status will be:
```
status:
  type: REGIONAL
  priority: 10
  conditions:
  - type: Enforced
     status: True
```

Reply:
If we are thinking about resource links, probably the following works:

```
status:
  type: VPCFirewall
  resource: global/firewalls/gke-pfw-default-my-ingress
  conditions:
  - type: Enforced
     status: True
```

For Global Network Firewall Policy:
```
status:
  type: GlobalNetworkFirewallPolicy
  resource: global/firewallPolicies/gke-pfw-default-my-ingress
  priority: 10
  conditions:
  - type: Enforced
     status: True
```

Does it look better than what is proposed right now?. I also wonder if we just want to keep the suffix part of resource link as this excludes GCP project info from the status. I feel including the full resource link might have some caveats (googleapis.com, compute alpha/beta/GA changes can modify the status)

Reply:
> With FW 3.0, firewall rules are not first class citizens in GCE anymore so they do not have self links. 

You mean FW 3.0 rule no longer has an object representation? I imagine all GCP api objects still has a resource URL right? 

> Also, the prefix googleapis.com keeps changing which might make the status uncertain.

This I agree. But we can counter that by having smarter util function for parsing.

Reply:
I am trying to point out that the Name and Global/Regional could not help pin point the actual GCP resource of FW rules.

Reply:
I agree with confusion about VPC/Global/Regional. Maybe we can change the names to VPCFirewall/GlobalNetworkFirewallPolicy/RegionNetworkFirewallPolicy.

Also, we will include the resource suffixes (excluding compute and project part).

FW 3.0 has single object for all the firewall rules. Within the object, priority uniquely identifies the rule. So, there is no resource link for a firewall rule but there is one for the network firewall policy.

```
status:
  type: GlobalNetworkFirewallPolicy
  resource: global/firewallPolicies/gke-pfw-default-my-ingress
  priority: 10
  conditions:
  - type: Enforced
     status: True
```

Does this look like something you are recommending?. In the above status, resource is same for all firewall rules, the priority uniquely identifies the rule.

Reply:
Updated according to our discussion today.